src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java

```diff
@@ -0,0 +1,44 @@
+package org.jenkinsci.plugins.workflow.pipelinegraphanalysis;
+
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.ChunkFinder;
+import org.jenkinsci.plugins.workflow.support.steps.StageStep;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * Finds both block-scoped and legacy stages
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public class StageChunkFinder implements ChunkFinder {
+
+    public boolean isStartInsideChunk() {
+        return true;
+    }
+
+    @Override
+    public boolean isChunkStart(@Nonnull FlowNode current, @CheckForNull FlowNode previous) {
+        LabelAction label = current.getAction(LabelAction.class);
+        return label != null && !(label instanceof ThreadNameAction);
```

Comment:
It was intended to improve performance slightly, but... I think at this point a premature optimization since there won't be that many nodes with a LabelAction.   You've convinced me, making the change.